File: mycollab-services-premium/src/main/java/com/mycollab/premium/license/service/LicenseResolverImpl.java
package com.mycollab.premium.license.service;

import com.mycollab.common.service.AppPropertiesService;
import com.mycollab.configuration.SiteConfiguration;
import com.mycollab.core.MyCollabException;
import com.mycollab.core.MyCollabVersion;
import com.mycollab.core.UserInvalidInputException;
import com.mycollab.core.utils.DateTimeUtils;
import com.mycollab.core.utils.FileUtils;
import com.mycollab.pro.license.LicenseInfo;
import com.mycollab.pro.license.LicenseType;
import com.verhas.licensor.License;
import org.joda.time.DateTime;
import org.joda.time.Duration;
import org.joda.time.LocalDate;
import org.joda.time.LocalDateTime;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.io.*;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.Properties;
import java.util.UUID;

/**
 * @author MyCollab Ltd
 * @since 5.2.6
 */
@Service
public class LicenseResolverImpl implements LicenseResolver, AppPropertiesService, InitializingBean {
    private static final Logger LOG = LoggerFactory.getLogger(LicenseResolverImpl.class);

    private Properties properties;
    private LicenseInfo licenseInfo = null;

    @Override
    public void afterPropertiesSet() throws Exception {
        File homeFolder = FileUtils.getHomeFolder();
        File sysFile = new File(homeFolder, ".app.properties");
        properties = new Properties();
        try {
            if (sysFile.isFile() && sysFile.exists()) {
                properties.load(new FileInputStream(sysFile));
                String startDate = properties.getProperty("startdate");
                if (startDate == null) {
                    properties.setProperty("startdate", DateTimeUtils.formatDateToW3C(new GregorianCalendar().getTime()));
                }
            } else {
                properties.setProperty("id", UUID.randomUUID().toString() + new LocalDateTime().getMillisOfSecond());
                properties.setProperty("startdate", DateTimeUtils.formatDateToW3C(new GregorianCalendar().getTime()));
            }
        } catch (IOException e) {
            LOG.error("Error", e);
        }

        File licenseFile = getLicenseFile();
        if (licenseFile.isFile() && licenseFile.exists()) {
            byte[] licenseBytes = org.apache.commons.io.FileUtils.readFileToByteArray(licenseFile);
            checkLicenseInfo(licenseBytes, false);
        } else {
            acquireALicense();
        }

        properties.store(new FileOutputStream(sysFile), "");
    }

    private File getLicenseFile() {
        File homeDir = FileUtils.getHomeFolder();
        File potentialOldFile = new File(homeDir, "mycollab.lic");
        File licenseFile = new File(homeDir, ".mycollab.lic");
        if (potentialOldFile.exists() && !licenseFile.exists()) {
            potentialOldFile.renameTo(licenseFile);
        }
        return licenseFile;
    }

    @Override
    public void checkAndSaveLicenseInfo(String licenseInputText) {
        try {
            checkLicenseInfo(licenseInputText.getBytes("UTF-8"), true);
        } catch (UnsupportedEncodingException e) {
            throw new MyCollabException("Error", e);
        }
    }

    private void acquireALicense() {
        LOG.info("Acquire the trial license");
        DateTime startDate = new DateTime(getStartDate());
        DateTime now = new DateTime();
        int days = (int) new Duration(startDate, now).getStandardDays();
        String edition = properties.getProperty("edition", "Community");
        if (!getEdition().equals(edition)) {
            properties.setProperty("edition", getEdition());
            RestTemplate restTemplate = new RestTemplate();
            try {
                String licenseRequest = restTemplate.postForObject(SiteConfiguration.getApiUrl("order/register-trial"), null, String.class);
                checkAndSaveLicenseInfo(licenseRequest);
            } catch (Exception e) {
                LOG.error("Can not retrieve a trial license", e);
                licenseInfo = createTempValidLicense(30 - days);
            }
        } else {
            licenseInfo = createInvalidLicense();
        }
    }

    @Override
    public LicenseInfo checkLicenseInfo(byte[] licenseBytes, boolean isSave) {
        try {
            License license = new License();
            InputStream publicStream = LicenseResolverImpl.class.getClassLoader().getResourceAsStream("pubring.gpg");
            license.loadKeyRing(publicStream, null);
            license.setLicenseEncoded(new ByteArrayInputStream(licenseBytes), "UTF-8");
            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
            license.dumpLicense(outputStream);
            outputStream.flush();
            Properties prop = new Properties();
            byte[] bytes = outputStream.toByteArray();
            prop.load(new ByteArrayInputStream(bytes));
            DateTime expireDate = new DateTime(DateTimeUtils.parseDateByW3C(prop.getProperty("expireDate")));
            if (MyCollabVersion.getReleasedDate() != null && MyCollabVersion.getReleasedDate().isAfter(expireDate)) {
                return createInvalidLicense();
            }
            LicenseInfo newLicenseInfo = new LicenseInfo();
            newLicenseInfo.setCustomerId(prop.getProperty("customerId"));
            newLicenseInfo.setLicenseType(LicenseType.valueOf(prop.getProperty("licenseType")));
            newLicenseInfo.setExpireDate(expireDate.toDate());
            newLicenseInfo.setIssueDate(DateTimeUtils.parseDateByW3C(prop.getProperty("issueDate")));
            newLicenseInfo.setLicenseOrg(prop.getProperty("licenseOrg"));
            newLicenseInfo.setMaxUsers(Integer.parseInt(prop.getProperty("maxUsers", "10")));
            if (isSave) {
                if (newLicenseInfo.isExpired()) {
                    throw new UserInvalidInputException("License is expired");
                }
                if (licenseInfo != null && newLicenseInfo.isTrial()) {
                    throw new UserInvalidInputException("You can not enter another trial license during trial period");
                }
                File licenseFile = getLicenseFile();
                FileOutputStream fileOutputStream = new FileOutputStream(licenseFile);
                fileOutputStream.write(licenseBytes);
                fileOutputStream.close();
            }
            licenseInfo = newLicenseInfo;
        } catch (Exception e) {
            licenseInfo = createInvalidLicense();
        }
        return licenseInfo;
    }

    private LicenseInfo createInvalidLicense() {
        LicenseInfo tmpLicenseInfo = new LicenseInfo();
        tmpLicenseInfo.setCustomerId("");
        tmpLicenseInfo.setExpireDate(new GregorianCalendar().getTime());
        tmpLicenseInfo.setIssueDate(new GregorianCalendar().getTime());
        tmpLicenseInfo.setLicenseOrg("");
        tmpLicenseInfo.setMaxUsers(1);
        tmpLicenseInfo.setLicenseType(LicenseType.INVALID);
        return tmpLicenseInfo;
    }

    private LicenseInfo createTempValidLicense(int days) {
        LocalDate now = new LocalDate();
        LicenseInfo tmpLicenseInfo = new LicenseInfo();
        tmpLicenseInfo.setCustomerId("");
        tmpLicenseInfo.setExpireDate(now.plusDays(days).toDate());
        tmpLicenseInfo.setIssueDate(now.toDate());
        tmpLicenseInfo.setLicenseOrg("");
        tmpLicenseInfo.setMaxUsers(10);
        tmpLicenseInfo.setLicenseType(LicenseType.PRO_TRIAL);
        return tmpLicenseInfo;
    }

    @Override
    public LicenseInfo getLicenseInfo() {
        return licenseInfo;
    }

    @Override
    public String getSysId() {
        return properties.getProperty("id", UUID.randomUUID().toString() + new LocalDateTime().getMillisOfSecond());
    }

    @Override
    public Date getStartDate() {
        try {
            String dateValue = properties.getProperty("startdate");
            return DateTimeUtils.convertDateByString(dateValue, "yyyy-MM-dd'T'HH:mm:ss");
        } catch (Exception e) {
            return new GregorianCalendar().getTime();
        }
    }

    @Override
    public String getEdition() {
        return "Premium";
    }
}

File: mycollab-license/src/main/java/com/mycollab/license/service/impl/LicenseResolverImpl.kt
package com.mycollab.license.service.impl

import com.javax0.license3j.licensor.License
import com.mycollab.common.service.AppPropertiesService
import com.mycollab.configuration.ServerConfiguration
import com.mycollab.core.MyCollabException
import com.mycollab.core.UserInvalidInputException
import com.mycollab.core.Version
import com.mycollab.core.utils.FileUtils
import com.mycollab.license.LicenseInfo
import com.mycollab.license.LicenseType
import com.mycollab.license.service.LicenseResolver
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.InitializingBean
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import java.io.*
import java.time.Duration
import java.time.LocalDate
import java.time.Period
import java.time.format.DateTimeFormatter
import java.util.*

/**
 * @author MyCollab Ltd
 * @since 5.2.6
 */
@Service
class LicenseResolverImpl(private val serverConfiguration: ServerConfiguration,
                          private val appPropertiesService: AppPropertiesService) : LicenseResolver, InitializingBean {

    val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")

    private lateinit var _licenseInfo: LicenseInfo
    override var licenseInfo: LicenseInfo? = null
        get() = _licenseInfo

    private val licenseFile: File
        get() {
            val homeDir = serverConfiguration.getHomeDir()
            val potentialOldFile = File(homeDir, "mycollab.lic")
            val licenseFile = File(homeDir, ".mycollab.lic")
            if (potentialOldFile.exists() && !licenseFile.exists()) {
                potentialOldFile.renameTo(licenseFile)
            }
            return licenseFile
        }

    @Throws(Exception::class)
    override fun afterPropertiesSet() {
        if (appPropertiesService.edition != "Cloud") {
            val licenseFile = licenseFile
            if (licenseFile.isFile && licenseFile.exists()) {
                val licenseBytes = org.apache.commons.io.FileUtils.readFileToByteArray(licenseFile)
                checkLicenseInfo(licenseBytes, false)
            } else {
                acquireALicense()
            }
        }
    }

    override fun checkAndSaveLicenseInfo(licenseInputText: String) {
        try {
            checkLicenseInfo(licenseInputText.toByteArray(charset("UTF-8")), true)
        } catch (e: UnsupportedEncodingException) {
            throw MyCollabException("Error", e)
        }

    }

    private fun acquireALicense() {
        LOG.info("Acquire the trial license")
        val startDate = LocalDate.from(appPropertiesService.startDate)
        val now = LocalDate.now()
        val days = Period.between(startDate, now).days
        val edition = appPropertiesService.edition
        val restTemplate = RestTemplate()
        try {
            val licenseRequest = restTemplate.postForObject(serverConfiguration.getApiUrl("order/register-trial"), null, String::class.java)
            checkAndSaveLicenseInfo(licenseRequest!!)
        } catch (e: Exception) {
            LOG.error("Can not retrieve a trial license", e)
            _licenseInfo = createTempValidLicense(30L - days)
        }
    }

    override fun checkLicenseInfo(licenseBytes: ByteArray, isSave: Boolean): LicenseInfo {
        try {
            val license = License()
            val publicStream = LicenseResolverImpl::class.java.classLoader.getResourceAsStream("pubring.gpg")
            license.loadKeyRing(publicStream, null)
            license.setLicenseEncoded(ByteArrayInputStream(licenseBytes), "UTF-8")
            val outputStream = ByteArrayOutputStream()
            license.dumpLicense(outputStream)
            outputStream.flush()
            val prop = Properties()
            val bytes = outputStream.toByteArray()
            prop.load(ByteArrayInputStream(bytes))
            val expireDate = LocalDate.parse(prop.getProperty("expireDate"), formatter)
            if (Version.getReleasedDate().isAfter(expireDate)) {
                return createInvalidLicense()
            }
            val newLicenseInfo = LicenseInfo()
            newLicenseInfo.customerId = prop.getProperty("customerId")
            newLicenseInfo.licenseType = LicenseType.valueOf(prop.getProperty("licenseType"))
            newLicenseInfo.expireDate = expireDate
            newLicenseInfo.issueDate = LocalDate.parse(prop.getProperty("issueDate"), formatter)
            newLicenseInfo.licenseOrg = prop.getProperty("licenseOrg")
            newLicenseInfo.maxUsers = Integer.parseInt(prop.getProperty("maxUsers", "10"))
            if (isSave) {
                if (newLicenseInfo.isExpired) {
                    throw UserInvalidInputException("License is expired")
                }
                if (newLicenseInfo.isTrial) {
                    throw UserInvalidInputException("You can not enter another trial license during trial period")
                }
                val licenseFile = licenseFile
                val fileOutputStream = FileOutputStream(licenseFile)
                fileOutputStream.write(licenseBytes)
                fileOutputStream.close()
            }
            _licenseInfo = newLicenseInfo
        } catch (e: Exception) {
            _licenseInfo = createInvalidLicense()
        }

        return _licenseInfo
    }

    private fun createInvalidLicense(): LicenseInfo {
        val tmpLicenseInfo = LicenseInfo()
        tmpLicenseInfo.customerId = ""
        tmpLicenseInfo.expireDate = LocalDate.now()
        tmpLicenseInfo.issueDate = LocalDate.now()
        tmpLicenseInfo.licenseOrg = ""
        tmpLicenseInfo.maxUsers = 1
        tmpLicenseInfo.licenseType = LicenseType.INVALID
        return tmpLicenseInfo
    }

    private fun createTempValidLicense(days: Long): LicenseInfo {
        val now = LocalDate.now()
        val tmpLicenseInfo = LicenseInfo()
        tmpLicenseInfo.customerId = ""
        tmpLicenseInfo.expireDate = now.plusDays(days)
        tmpLicenseInfo.issueDate = now
        tmpLicenseInfo.licenseOrg = ""
        tmpLicenseInfo.maxUsers = 10
        tmpLicenseInfo.licenseType = LicenseType.PRO_TRIAL
        return tmpLicenseInfo
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(LicenseResolverImpl::class.java)
    }
}
